esphome/components/remote_base/symphony_protocol.cpp

```diff
@@ -0,0 +1,118 @@
+#include "symphony_protocol.h"
+#include "esphome/core/log.h"
+
+namespace esphome {
+namespace remote_base {
+
+static const char *const TAG = "remote.symphony";
+
+// Reference implementation and timing details:
+// IRremoteESP8266 ir_Symphony.cpp
+// https://github.com/crankyoldgit/IRremoteESP8266/blob/master/src/ir_Symphony.cpp
+// The implementation below mirrors the constant bit-time mapping and
+// footer-gap handling used there.
+
+// Symphony protocol timing specifications (tuned to handset captures)
+static const uint32_t BIT_ZERO_HIGH_US = 460;  // short
+static const uint32_t BIT_ZERO_LOW_US = 1260;  // long
+static const uint32_t BIT_ONE_HIGH_US = 1260;  // long
+static const uint32_t BIT_ONE_LOW_US = 460;    // short
+static const uint32_t CARRIER_FREQUENCY = 38000;
+
+// IRremoteESP8266 reference: kSymphonyFooterGap = 4 * (mark + space)
+static const uint32_t FOOTER_GAP_US = 4 * (BIT_ZERO_HIGH_US + BIT_ZERO_LOW_US);
+// Typical inter-frame gap (~34.8 ms observed)
+static const uint32_t INTER_FRAME_GAP_US = 34760;
+
+void SymphonyProtocol::encode(RemoteTransmitData *dst, const SymphonyData &data) {
+  dst->set_carrier_frequency(CARRIER_FREQUENCY);
+  ESP_LOGD(TAG, "Sending Symphony: data=0x%0*X nbits=%u repeats=%u", (data.nbits + 3) / 4, (unsigned int) data.data,
+           data.nbits, data.repeats);
+  // Each bit produces a mark+space (2 entries). We fold the inter-frame/footer gap
+  // into the last bit's space of each frame to avoid over-length gaps.
+  dst->reserve(data.nbits * 2u * data.repeats);
+
+  for (uint8_t repeats = 0; repeats < data.repeats; repeats++) {
+    // Data bits (MSB first)
+    for (uint32_t mask = 1UL << (data.nbits - 1); mask != 0; mask >>= 1) {
+      const bool is_last_bit = (mask == 1);
+      const bool is_last_frame = (repeats == (data.repeats - 1));
+      if (is_last_bit) {
+        // Emit last bit's mark; replace its space with the proper gap
+        if (data.data & mask) {
+          dst->mark(BIT_ONE_HIGH_US);
+        } else {
+          dst->mark(BIT_ZERO_HIGH_US);
+        }
+        dst->space(is_last_frame ? FOOTER_GAP_US : INTER_FRAME_GAP_US);
+      } else {
+        if (data.data & mask) {
+          dst->item(BIT_ONE_HIGH_US, BIT_ONE_LOW_US);
+        } else {
+          dst->item(BIT_ZERO_HIGH_US, BIT_ZERO_LOW_US);
+        }
+      }
+    }
+  }
+}
+
+optional<SymphonyData> SymphonyProtocol::decode(RemoteReceiveData src) {
+  auto is_valid_len = [](uint8_t nbits) -> bool { return nbits == 8 || nbits == 12 || nbits == 16; };
+
+  RemoteReceiveData s = src;  // copy
+  SymphonyData out{0, 0, 1};
+
+  for (; out.nbits < 32; out.nbits++) {
+    uint32_t bit;
+    if (s.expect_mark(BIT_ONE_HIGH_US)) {
+      bit = 1;
```

Comment:
Why not just do `out.data = (out.data << 1UL) | 1;` here
```suggestion
      out.data = (out.data << 1UL) | 1UL;
```

Reply:
> Why not just do `out.data = (out.data << 1UL) | 1;` here

It's a good suggestion. But I need to remove the common update at the end, otherwise, in bit 1, it would update twice in the same iteration.

And use the update `out.data` inside both branches:
For bit = 1: `(out.data << 1UL) | 1UL`
For bit = 0: `(out.data << 1UL) | 0UL`

I used `continue;` after each inline update to make the flow explicit, although it's optional